test(docs-infra): add unit tests for rxjs examples

sonukapoor opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [X] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

sonukapoor committed:
test(docs-infra): add unit tests for rxjs examples

This commit adds missing unit tests for all rxjs examples from the docs.

Closes #28017

sonukapoor commented:
@gkalpak Please review draft.

gkalpak commented:
This lgtm for now. We will re-work it the infrastructure, before the merging the PR, but this is good enough for now so that you can keep working on the actual tests. Thx for sorting it out 👍

gkalpak commented:
As discussed "offline", here are my thoughts (and suggestions for next steps):

- A proper solution might involve setting up a different project type for Observable-related projects (e.g. not involving Angular dependencies and providing dependencies needed for observables only). This would be a little expensive in terms of both initial implementation and maintenance, though.

- I realized that the observables examples are not offered for download/StackBlitz, so the users only really see the code snippets (not the whole files). This gives us some flexibility on putting stuff around the docregions that will help us with testing, but without confusing the users (which will never see that extra stuff).

- I thought about using Karma for the tests. This would give us access to a real browser. I then realized that that would only solve part of the problem. We would need to add appropriate HTML files and that would take more time/maintenance/complexity and would most likely still require infrastructure changes. Not ideal.

- Finally, I noticed another issue: The code in the files executes eagerly (i.e. as soon as the file is loaded), which means that we would have a hard time testing stuff in the test files, because it would have happened before even loading the test files.

---
So, given the above, I think a good solution ("good" in the sense that it is fairly simple, will not take a lot of time to implement infrastructure for and will allow us to focus on tests (which is the primary focus for now)) would be the following:

- We could wrap each docregion (i.e. code snippet) in a function and export that function. This way, the code snippets that the user sees wouldn't change (because they would only see the function body), but the code would not be executed as soon as the file is loaded during tests.

- In tests, we could call the exported functions to run the code. This gives us the freedom to:
  a. Set up spies and mocks before executing the function.
  b. Pass mocks for globals as arguments to the function. For example, we could pass a mock `document` to the function etc.

As an example, here is what a code snippet could look like before my suggested change:

```js
// #docregion foo
const elem = document.querySelector('foo');
const subscription = fromEvent(elem, 'click').subscribe(() => console.log('*click*'));

setTimeout(() => subscription.unsubscribe(), 10000);
// #enddocregion foo
```

And here is how it would be transformed with my suggestion:

```js
// By wrapping a function around the code, we make it "lazy"
// (i.e. it won't be executed immediately).
//
// We can pass in "globals" (in TS we can give them proper types).
// Here for example, we can pass in a mock `console` to capture the logs.
// (We could even pass a mock `setTimeout()`, but that is unnecessary,
// since we can mock that with `jasmine.clock().install()` etc.)
export function docregionFoo(console, document) {

  // The code-snippet that the user will see remains unchanged.
  // (The indentation might change, but that is fine.)
  // #docregion foo
  const elem = document.querySelector('foo');
  const subscription = fromEvent(elem, 'click').subscribe(() => console.log('*click*'));

  setTimeout(() => subscription.unsubscribe(), 10000);
  // #enddocregion foo

  // Here, return the `subscription`, if we need to run assertions against it.
  return subscription;
}
```

petebacondarwin commented:
@sonukapoor can we close this PR?

sonukapoor commented:
> @sonukapoor can we close this PR?

This PR is awaiting @gkalpak's review :)

sonukapoor commented:
Blocked by https://github.com/angular/angular/pull/38143

sonukapoor commented:
@gkalpak This is ready for you to review. Thanks

gkalpak commented on aio/content/examples/observables/src/multicasting.ts:
```diff
@@ -23,17 +24,31 @@ function sequenceSubscriber(observer) {
+  we need to start the indentation of the below body function at the most left
+  position.
+*/
+export function docDelaySequence(console) {
```
```suggestion
export function docRegionDelaySequence(console) {
```

gkalpak commented on aio/content/examples/observables/src/multicasting.ts:
```diff
@@ -23,17 +24,31 @@ function sequenceSubscriber(observer) {
 sequence.subscribe({
-  next(num) { console.log(num); },
-  complete() { console.log('Finished sequence'); }
+  next(num) {
+    console.log(num);
+  },
+  complete() {
+    console.log('Finished sequence');
+  }
```
These changes look unrelated. Please revert.

gkalpak commented on aio/content/examples/observables/src/multicasting.ts:
```diff
@@ -48,15 +63,23 @@ sequence.subscribe({
 sequence.subscribe({
-  next(num) { console.log('1st subscribe: ' + num); },
-  complete() { console.log('1st sequence finished.'); }
+  next(num) {
+    console.log('1st subscribe: ' + num);
+  },
+  complete() {
+    console.log('1st sequence finished.');
+  }
```
These changes look unrelated. Please revert.

gkalpak commented on aio/content/examples/observables/src/multicasting.ts:
```diff
@@ -48,15 +63,23 @@ sequence.subscribe({
   sequence.subscribe({
-    next(num) { console.log('2nd subscribe: ' + num); },
-    complete() { console.log('2nd sequence finished.'); }
+    next(num) {
+      console.log('2nd subscribe: ' + num);
+    },
+    complete() {
+      console.log('2nd sequence finished.');
+    }
```
These changes look unrelated. Please revert.

gkalpak commented on aio/content/examples/observables/src/multicasting.ts:
```diff
@@ -84,20 +109,24 @@ function multicastSequenceSubscriber() {
-      }, seq, 0);
+        seq,
+        0
+      );
```
Same as above. Please revert the unrelated, formatting changes.

gkalpak commented on aio/content/examples/observables/src/multicasting.ts:
```diff
@@ -126,6 +155,12 @@ function doSequence(observer, arr, idx) {
+// Because of how the code is merged together using the doc regions,
+// we need to start the indention of the below body function at the most left
+// position.
+export function docMulticastSequence(console) {
```
```suggestion
export function docRegionMulticastSequence(console) {
```

gkalpak commented on aio/content/examples/observables/src/multicasting.spec.ts:
```diff
@@ -0,0 +1,27 @@
+  let console;
+  beforeEach(() => {
+    console = {
+      log: jasmine.createSpy
```
Was this supposed to be `jasmine.createSpy('log')`?

gkalpak commented on aio/content/examples/observables/src/multicasting.spec.ts:
```diff
@@ -0,0 +1,27 @@
+      expect(output).toEqual([1, 2, 3]);
+      doneFn();
+    });
+  });
```
Same as above, this test does not verify what happens with the subscriptions inside `docDelaySequence()`. This is the main thing we need to assert. (Same for the test below.)

gkalpak reviewed (changes requested):
Here is my partial review. Many of the comments apply to files in all observables examples, so please make the changes to all of them.

sonukapoor commented on aio/content/examples/observables/src/creating.ts:
```diff
@@ -16,6 +17,14 @@ function sequenceSubscriber(observer) {
+  Because of how the code is merged together using the doc regions,
+  we need to start the indentation of the below body function at the most left
+  position.
+*/
```
Hm... I am pretty sure, I had tried this and it was not indenting properly, which is why I had the comment in there in the first place. I have tried this again now and it seems to work fine 👍

sonukapoor commented on aio/content/examples/observables/src/creating.spec.ts:
```diff
@@ -0,0 +1,49 @@
+  it('should create an observable using the constructor', () => {
+    const console = { log: jasmine.createSpy };
+    const source = docRegionSubscriber(console);
+    const expected = cold('(abc|)', { a: 1, b: 2, c: 3 });
```
There is only one observable. Yes, It will subscribe to the emitted values and verify them. 
Ok. I will remove the marbles and verify that `console.log` was called instead.

sonukapoor commented on aio/content/examples/observables/src/creating.ts:
```diff
@@ -51,16 +65,19 @@ function fromEvent(target, eventName) {
-const subscription = fromEvent(nameInput, 'keydown')
-  .subscribe((e: KeyboardEvent) => {
-    if (e.keyCode === ESC_KEY) {
+  const subscription = fromEvent(nameInput, 'keydown');
+  subscription.subscribe((e: KeyboardEvent) => {
```
I am using the `subscription` to make sure that the `removeEventListener` is called when unsubscribing. See the test [here](https://github.com/sonukapoor/angular/blob/unit-tests-for-stand-alone-code/aio/content/examples/observables/src/creating.spec.ts#L32).

gkalpak replied:
I don't think you need this change to test the unsubscription. You can just change [this line][1] of the test:

```diff
-    subscription.subscribe().unsubscribe();
+    subscription.unsubscribe();

[1]: https://github.com/sonukapoor/angular/blob/230b42f1bfeee39d40538c443533d6e876d2faad/aio/content/examples/observables/src/creating.spec.ts#L53

sonukapoor replied:
Ok. Would you like me to remove the test?

sonukapoor commented:
@gkalpak, I have addressed your review comments. However, I have some difficulties to get the doc-plasters working without the indentation when there are imports used. Please see the individual sections where I have added a screenshot with the issue.

mary-poppins commented:
You can preview 48592a0 at https://pr34190-48592a0.ngbuilds.io/.
You can preview b1fcd71 at https://pr34190-b1fcd71.ngbuilds.io/.
You can preview dff4783 at https://pr34190-dff4783.ngbuilds.io/.
You can preview fe5d2cd at https://pr34190-fe5d2cd.ngbuilds.io/.
You can preview e4e23a4 at https://pr34190-e4e23a4.ngbuilds.io/.
You can preview ced4457 at https://pr34190-ced4457.ngbuilds.io/.
You can preview 6f51f80 at https://pr34190-6f51f80.ngbuilds.io/.
You can preview a60f18c at https://pr34190-a60f18c.ngbuilds.io/.
You can preview 389ba58 at https://pr34190-389ba58.ngbuilds.io/.
You can preview a8d8017 at https://pr34190-a8d8017.ngbuilds.io/.

mary-poppins commented:
You can preview f40a371 at https://pr34190-f40a371.ngbuilds.io/.

mary-poppins commented:
You can preview 230b42f at https://pr34190-230b42f.ngbuilds.io/.

gkalpak commented on aio/content/examples/observables/src/creating.ts:
```diff
@@ -2,51 +2,38 @@
+    // because values are delivered synchronously
+    return {unsubscribe() {}};
+  }
+  // Create a new Observable that will deliver the above sequence
```
```suggestion

  // Create a new Observable that will deliver the above sequence
```

gkalpak commented on aio/content/examples/observables/src/creating.spec.ts:
```diff
@@ -1,12 +1,16 @@
 describe('observables', () => {
   it('should create an observable using the constructor', () => {
-    const console = { log: jasmine.createSpy };
+    const consoleLogSpy  = spyOn(console , 'log');
```
Generally, it is preferable to use a spy object (instead of spying on the real `console.log()` (in case some other pieve of code wants to `console.log()` something):

```suggestion
    const console = {log: jasmine.createSpy('log')};
```

gkalpak commented on aio/content/examples/observables/src/creating.spec.ts:
```diff
@@ -1,12 +1,16 @@
     const source = docRegionSubscriber(console);
-    const expected = cold('(abc|)', { a: 1, b: 2, c: 3 });
-    expect(source).toBeObservable(expected);
+    source.subscribe(() => {
```
Is subscribing here necessary?
Couldn't we run the assertions immediately?

gkalpak commented on aio/content/examples/practical-observable-usage/src/backoff.spec.ts:
```diff
@@ -0,0 +1,31 @@
+import { backoff } from './backoff';
+import { map } from 'rxjs/operators';
+import { Subject } from 'rxjs';
```
```suggestion
import { Subject } from 'rxjs';
import { map } from 'rxjs/operators';
import { backoff } from './backoff';
```

gkalpak commented on aio/content/examples/practical-observable-usage/src/backoff.spec.ts:
```diff
@@ -0,0 +1,31 @@
+        // should not emit 4
+        expect(a).not.toEqual(4);
+        if (a === 3) {
+          doneFn();
```
This test looks broken (for a couple fo reasons).

1. How are we sure 4 is not emitted if we exit on 3? :grin:
2. Since 4 is emitted once (i.e. the observable only errors once), how are we verifying the backoff?

gkalpak commented on aio/content/examples/practical-observable-usage/src/backoff.spec.ts:
```diff
@@ -0,0 +1,31 @@
+    obs.next(2);
+    obs.next(3);
+    obs.next(4);
+  });
```
Ideally, we should also test the backing off (i.e. the increased delay between retries).
Since this is a little complicated, feel free to add a TODO comment for us to do later (at some point...maybe 😛).

gkalpak commented on aio/content/examples/practical-observable-usage/src/typeahead.ts:
```diff
@@ -1,18 +1,32 @@
+*/
+// #docplaster
+// #docregion
 
```
```suggestion
```

gkalpak commented on aio/content/examples/practical-observable-usage/src/typeahead.spec.ts:
```diff
@@ -0,0 +1,29 @@
+import { docRegionTypeahead } from './typeahead';
+import { of } from 'rxjs';
```
```suggestion
import { of } from 'rxjs';
import { docRegionTypeahead } from './typeahead';
```

gkalpak commented on aio/content/examples/observables/src/multicasting.spec.ts:
```diff
@@ -0,0 +1,45 @@
+import { docRegionDelaySequence, docRegionMulticastSequence } from './multicasting';
+import { toArray } from 'rxjs/operators';
```
```suggestion
import { toArray } from 'rxjs/operators';
import { docRegionDelaySequence, docRegionMulticastSequence } from './multicasting';
```

gkalpak commented on aio/content/examples/observables/src/subscribing.spec.ts:
```diff
@@ -0,0 +1,22 @@
+import { docRegionObserver } from './subscribing';
+import { toArray } from 'rxjs/operators';
```
```suggestion
import { toArray } from 'rxjs/operators';
import { docRegionObserver } from './subscribing';
```

gkalpak commented on aio/content/examples/practical-observable-usage/src/typeahead.spec.ts:
```diff
@@ -0,0 +1,29 @@
+    });
+
+    triggertInputChange({ target: { value: 'Hello' } });
+  });
```
This test changes too little of the implementation. Even though testing everything is quite complicated (and thus outside the scope of this PR), there is some low-hanging fruit:

- Make the mock `ajax` a spy so we can track calls.
- Verify that there is no AJAX call when the value is less than 3 characters.
- Verify that there is not AJAX call when the value of the input does not change (even if the `input` event is emitted on the element).
- Potentially, use jasmine's [mock clock][1] (or RxJS' [TestScheduler][2]?) to verify that events are debounced.

##
**Bonus points:**

The current example that is calling the same API endpoint without taking the emitted value into account is rather unrealistic.

- Improve the example, by using the emitted input value, for example:
  ```diff
       distinctUntilChanged(),
  -    switchMap(() => ajax('/api/endpoint'))
  +    switchMap(searchTerm => ajax(`/api/endpoint?search=${searchTerm}`))
     );
  ```
2. Update the test (or add a new test) to verify that AJAX calls are made to the correct endpoint.
3. Update the test (or add a new test) to verify that no AJAX calls are made for intermediate values when debouncing.

[1]: https://jasmine.github.io/api/3.6/Clock
[2]: https://rxjs.dev/api/testing/TestScheduler

sonukapoor replied:
I have added the scenarios into the same commit. But let me know if you want me to split this up into a new commit (which might make sense).

gkalpak reviewed (changes requested):
Thx, @sonukapoor! Here's another round of review comments :smiley:
(I have reviewed all examples except for the `rx-library` one.)

Pending comments from previous review (for my reference):
- https://github.com/angular/angular/pull/34190#discussion_r482845609

mary-poppins commented:
You can preview e7bc538 at https://pr34190-e7bc538.ngbuilds.io/.

sonukapoor commented on aio/content/examples/observables/src/multicasting.spec.ts:
```diff
@@ -1,26 +1,44 @@
   it('should create an observable and emit in sequence', (doneFn: DoneFn) => {
-    const sequence = docDelaySequence(console);
+    const sequence = docRegionDelaySequence(console);
     sequence.pipe(toArray()).subscribe((output: Array<number>) => {
```
I have changed all tests to not subscribe anymore.

gkalpak reviewed (changes requested):
We are getting there :smiley:

In addition to my other comments, can you also:
- Move the change of using `searchTerm` in the typeahead request to a separate commit (as discussed in https://github.com/angular/angular/pull/34190#discussion_r484560207) and adjust the tests accordingly (i.e. each commit should include the tests that are relevant to the source code changes it introduces).
- Move the changes to the `rx-library` example to a new PR (in the interest of getting this one in faster and avoiding its getting stale).

sonukapoor committed:
test(docs-infra): improve typeahead example and add unit test

This commit improves the typeahead example, by using the emitted input
value. It also adds a unit test to ensure that the example is working
as intended.

sonukapoor commented on aio/content/examples/practical-observable-usage/src/backoff.spec.ts:
```diff
@@ -0,0 +1,31 @@
+import { backoff } from './backoff';
+import { map } from 'rxjs/operators';
+import { Subject } from 'rxjs';
```
I have added a `TODO` to fix the test, once this PR is merged.

sonukapoor commented on aio/content/examples/practical-observable-usage/src/typeahead.spec.ts:
```diff
@@ -0,0 +1,64 @@
+    expect(ajax).not.toHaveBeenCalled();
+  });
+
+  it('should make an ajax call to the corrent endpoint', () => {
```
Done.

sonukapoor commented on aio/content/examples/observables/src/multicasting.spec.ts:
```diff
@@ -0,0 +1,44 @@
+    expect(console.log).toHaveBeenCalledWith('1st subscribe: 3');
+    expect(console.log).toHaveBeenCalledWith('1st sequence finished.');
+    expect(console.log).toHaveBeenCalledWith('2nd subscribe: 3');
+    expect(console.log).toHaveBeenCalledWith('2nd sequence finished.');
```
Yes, that makes it a lot easier 👍

gkalpak commented on aio/content/examples/practical-observable-usage/src/typeahead.ts:
```diff
@@ -1,18 +1,32 @@
 
+// #enddocregion
+/* tslint:disable:no-shadowed-variable */
+/* tslint:disable:align */
```
Is this needed? Why?

sonukapoor replied:
> Is this needed? Why?

Because of the alignment of the imports. Because we are hiding the function declaration using the doc regions the import statements are misaligned with the body of the function. In the below example, the `import` aligns perfectly fine with the body `const bar...`.

```typescript
// #docplaster
// #docregion
  import {} ...
// #enddocregion
export function foo() {
  // #docRegion
  const bar = ...
  // #enddocregion
}
```

gkalpak reviewed (approved):
Almost ready :tada:

BTW, please update the second commit message to use imperative, present tesne (per our [guidelines](https://github.com/angular/angular/blob/1d9e56eba06958e37ef8062ae7b65afa85f90632/CONTRIBUTING.md#commit)):

`improves typeahead example and adds unit test` --> `improve typeahead example and add unit test …`

gkalpak reviewed (approved):
Nice :tada:

Reviewed-for: global-docs-approvers
